webapp/src/components/common/GroupedDataTable/index.tsx

```diff
@@ -1,69 +1,299 @@
+import type { TRow } from "./types";
 
-export interface GroupedDataTableProps<TData extends TRow> {
+export interface GroupedDataTableProps<
+  TGroups extends string[],
+  TData extends TRow<TGroups[number]>,
+> {
   data: TData[];
-  columns: Array<MRT_ColumnDef<TData>>;
-  groups: string[] | readonly string[];
-  onCreate?: (values: TData) => Promise<TData>;
-  onDelete?: (ids: string[]) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  columns: Array<MRT_ColumnDef<TData, any>>;
+  groups: TGroups;
+  onCreate?: (values: TRow<TGroups[number]>) => Promise<TData>;
+  onDuplicate?: (row: TData, newName: string) => Promise<TData>;
+  onDelete?: (rows: TData[]) => PromiseAny | void;
+  onNameClick?: (row: TData) => void;
+  isLoading?: boolean;
+  deleteConfirmationMessage?: string | ((count: number) => string);
+  fillPendingRow?: (
+    pendingRow: TRow<TGroups[number]>,
+  ) => TRow<TGroups[number]> & Partial<TData>;
 }
 
-function GroupedDataTable<TData extends TRow>({
+// Use ids to identify default columns (instead of `accessorKey`),
+// to have a unique identifier. It is more likely to have a duplicate
+// `accessorKey` with `columns` prop.
+const GROUP_COLUMN_ID = "_group";
+const NAME_COLUMN_ID = "_name";
+
+function GroupedDataTable<
+  TGroups extends string[],
+  TData extends TRow<TGroups[number]>,
+>({
   data,
   columns,
   groups,
   onCreate,
+  onDuplicate,
   onDelete,
-}: GroupedDataTableProps<TData>) {
+  onNameClick,
+  isLoading,
+  deleteConfirmationMessage,
+  fillPendingRow,
+}: GroupedDataTableProps<TGroups, TData>) {
   const { t } = useTranslation();
   const [openDialog, setOpenDialog] = useState<
     "add" | "duplicate" | "delete" | ""
   >("");
   const [tableData, setTableData] = useState(data);
   const [rowSelection, setRowSelection] = useState<MRT_RowSelectionState>({});
+  const enqueueErrorSnackbar = useEnqueueErrorSnackbar();
+  const callbacksRef = useAutoUpdateRef({ onNameClick });
+  const prevData = usePrevious(data);
+  const pendingRows = useRef<Array<TRow<TGroups[number]>>>([]);
+  const { createOps, deleteOps, totalOps } = useOperationInProgressCount();
 
-  const isAnyRowSelected = useMemo(
-    () => Object.values(rowSelection).some((value) => value),
-    [rowSelection],
-  );
-
-  const isOneRowSelected = useMemo(
-    () => Object.values(rowSelection).filter((value) => value).length === 1,
-    [rowSelection],
-  );
-
-  const selectedRow = useMemo(() => {
-    if (isOneRowSelected) {
-      const selectedIndex = Object.keys(rowSelection).find(
-        (key) => rowSelection[key],
-      );
-      return selectedIndex && tableData[+selectedIndex];
+  // Update once `data` only if previous value was empty.
+  // It allows to handle loading data.
+  useUpdateEffectOnce(() => {
+    if (prevData && prevData.length === 0) {
+      setTableData(data);
     }
-  }, [isOneRowSelected, rowSelection, tableData]);
+  }, [data.length]);
```

Comment:
I dont know what was your intent here but this is causing a major state bug, the tableData is never updated, so at each re-render even if we change the area the state remains the same. here's my analysis:

- initial fetch: when data changes from empty to filled, `prevData` is empty `([])`, and data is now filled `([...items])`. The effect runs, setting tableData with data. This is the first and only time this effect runs.

- data changes: if data changes again (like we change the area), useUpdateEffectOnce will not execute its effect function again because `hasUpdated.current` is permanently set to `true` after the first execution. no updates to `tableData` will occur no matter how data changes.

Reply:
Thanks for testing, I will investigate.

Reply:
Before `UsePromiseCond` prevent this issue (`GroupedDataTable` was unmounted but now no), but I remove it, so I need to add a `key` in `GroupedDataTable`.

Reply:
Ok so finally it's not because of the `useEffectOnce` 🤔  thanks for the return I'll take a look 👍🏼